zetaclient/chains/bitcoin/observer/inbound.go

```diff
@@ -472,3 +472,173 @@ func GetBtcEvent(
 	}
 	return nil, nil
 }
+
+// GetBtcEventWithWitness either returns a valid BTCInboundEvent or nil.
+// This method supports data with more than 80 bytes by scanning the witness for possible presence of a tapscript.
+// It will first prioritize OP_RETURN over tapscript.
+func GetBtcEventWithWitness(
+	client interfaces.BTCRPCClient,
+	tx btcjson.TxRawResult,
+	tssAddress string,
+	blockNumber uint64,
+	logger zerolog.Logger,
+	netParams *chaincfg.Params,
+	depositorFee float64,
+) (*BTCInboundEvent, error) {
+	if len(tx.Vout) < 1 {
+		logger.Debug().Msgf("no output %s", tx.Txid)
+		return nil, nil
+	}
+	if len(tx.Vin) == 0 { // should never happen
+		logger.Debug().Msgf("no input found for inbound: %s", tx.Txid)
+		return nil, nil
+	}
+
+	if !isValidRecipient(tx.Vout[0].ScriptPubKey.Hex, tssAddress, netParams, logger) {
+		logger.Debug().Msgf("irrelevant recipient %s for inbound: %s", tx.Vout[0].ScriptPubKey.Hex, tx.Txid)
+		return nil, nil
+	}
```

Comment:
I think we could have a cleaner api on this case that avoids repeating the same logging information.

We currently logs the errors in `isValidRecipient`, then here, and then we return nil.

Instead of this approach:
- First, `isValidRecipient` should return an error instead of a bool. Modify the returned errors to include the address as well.
- Then we can proceed to modify this bit:
```suggestion
	if err := isValidRecipient(tx.Vout[0].ScriptPubKey.Hex, tssAddress, netParams, logger); err != nil {
		logger.Warn().Msgf("unexpected inbound %s: %s", tx.Txid, err)
		return nil, nil
	}
```

Reply:
I dont think it should be necessary to return an error if invalid recipient is detected. This is not an error, it's just not for zeta, pretty normal situation. I think perhaps what's worth more special treatment is when the recipient is correct, but other things go wrong, like invalid script or invalid amount. Maybe for data analytics or UX alert. Maybe.